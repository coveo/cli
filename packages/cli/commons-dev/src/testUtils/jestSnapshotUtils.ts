// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string) =>
  input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/gm, '›');

export const formatAbsolutePath = (input: string) =>
  input.replaceAll(/\w:\\{2}/gm, '/').replaceAll(/\\{2}/gm, '/');

export const removeDotCmd = (input: string) => input.replaceAll(/\.cmd/gm, '');
