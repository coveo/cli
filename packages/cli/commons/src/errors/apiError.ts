import {red} from 'chalk';
import dedent from 'ts-dedent';
import {PrintableError, SeverityLevel} from './printableError';

export interface APIErrorResponse {
  message: string;
  errorCode: string;
  requestID: string;
}

export interface AxiosErrorFromAPI {
  response: {
    status: number;
    data: APIErrorResponse;
  };
}

export const APIErrorSchema = {
  type: 'object',
  required: ['errorCode', 'message'],
  properties: {
    errorCode: {
      type: 'string',
    },
    message: {
      type: 'string',
    },
  },
};

export const AxiosErrorFromAPISchema = {
  type: 'object',
  required: ['response'],
  properties: {
    response: {
      type: 'object',
      required: ['status', 'data'],
      properties: {
        status: {
          type: 'integer',
        },
        data: APIErrorSchema,
      },
    },
  },
};

export class APIError extends PrintableError {
  public constructor(
    error: APIErrorResponse | AxiosErrorFromAPI,
    tagLine?: string
  ) {
    super(SeverityLevel.Error);
    let errorCode: string;
    let message: string;
    let status: number | undefined;
    if (this.isFromAxios(error)) {
      errorCode = error.response.data.errorCode;
      message = error.response.data.message;
      status = error.response.status;
    } else {
      errorCode = error.errorCode;
      message = error.message;
    }

    this.name = `APIError - ${errorCode}`;
    this.message = dedent`
    ${tagLine}
    ${status ? `Status code: ${status}\n` : ''}
    Error code: ${red(errorCode)}
    Message: ${red(message)}
    `;
  }

  private isFromAxios(
    error: APIErrorResponse | AxiosErrorFromAPI
  ): error is AxiosErrorFromAPI {
    return 'response' in error;
  }
}
