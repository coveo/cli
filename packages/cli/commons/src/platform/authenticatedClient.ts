import {setGlobalDispatcher, ProxyAgent, FormData, fetch} from 'undici';
import PlatformClient from '@coveo/platform-client';
import {Config, Configuration} from '../config/config';
import {castEnvironmentToPlatformClient} from './environment';
import globalConfig from '../config/globalConfig';
import {lt} from 'semver';

export class AuthenticatedClient {
  public cfg: Config;
  public constructor() {
    this.cfg = new Config(globalConfig.get().configDir);
  }

  public async isLoggedIn() {
    const {accessToken} = this.cfg.get();
    return accessToken !== undefined;
  }

  public async isExpired() {
    try {
      const c = await this.getClient();
      await c.initialize();
      return false;
    } catch (e) {
      return true;
    }
  }

  public async getClient(overrideConfig?: Partial<Configuration>) {
    const configFromDisk = this.cfg.get();
    const resolvedConfig = {...configFromDisk, ...overrideConfig};
    const globalRequestSettings: Record<string, unknown> = {};
    const proxyServer =
      process.env['https_proxy'] || process.env['HTTPS_PROXY'];
    if (proxyServer) {
      setGlobalDispatcher(new ProxyAgent(proxyServer));
    }
    if (lt(process.version, '18.0.0')) {
      Object.assign(global, {FormData, fetch});
    }
    return new PlatformClient({
      globalRequestSettings,
      environment: castEnvironmentToPlatformClient(resolvedConfig.environment),
      region: resolvedConfig.region,
      organizationId: resolvedConfig.organization,
      accessToken: resolvedConfig.accessToken!,
    });
  }

  public async getAllOrgsUserHasAccessTo() {
    const platformClient = await this.getClient();
    return platformClient.organization.list();
  }

  public async getUserHasAccessToOrg(org: string) {
    const orgs = await this.getAllOrgsUserHasAccessTo();
    return orgs.some((o) => o.id === org);
  }

  public async createImpersonateApiKey(name: string) {
    const platformClient = await this.getClient();
    return await platformClient.apiKey.create({
      displayName: `cli-${name}`,
      description: 'Generated by the Coveo CLI',
      enabled: true,
      privileges: [
        {targetDomain: 'IMPERSONATE', targetId: '*', owner: 'SEARCH_API'},
      ],
    });
  }

  public async getUsername() {
    const {anonymous, organization} = this.cfg.get();
    if (anonymous) {
      return `api-key@${organization}`;
    }
    const authenticatedClient = new AuthenticatedClient();
    const platformClient = await authenticatedClient.getClient();
    await platformClient.initialize();

    return (await platformClient.user.get()).providerUsername;
  }
}

export enum AuthenticationStatus {
  LOGGED_IN,
  EXPIRED,
  LOGGED_OUT,
}

export async function getAuthenticationStatus() {
  const authenticatedClient = new AuthenticatedClient();

  if (!(await authenticatedClient.isLoggedIn())) {
    return AuthenticationStatus.LOGGED_OUT;
  }

  if (await authenticatedClient.isExpired()) {
    return AuthenticationStatus.EXPIRED;
  }

  return AuthenticationStatus.LOGGED_IN;
}
