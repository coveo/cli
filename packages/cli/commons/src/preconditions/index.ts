export * from './authenticated';
export * from './apiKeyPrivilege';
export * from './preconditions';
