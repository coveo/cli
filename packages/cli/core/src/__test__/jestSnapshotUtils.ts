import strip from 'strip-ansi';

// See https://github.com/oclif/core/blob/f996086397158b5beafc4254718c78fd93d23b1b/src/errors/errors/cli.ts#L55-L60
export const formatCliLog = (input: string, stripAinsi = false) => {
  const formatted = input.replaceAll(/\r\n/gm, '\n').replaceAll(/»/g, '›');
  return stripAinsi ? strip(formatted) : formatted;
};
