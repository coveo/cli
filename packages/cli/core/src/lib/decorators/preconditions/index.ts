// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
