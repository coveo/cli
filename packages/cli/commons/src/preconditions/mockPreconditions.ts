import {CLICommand} from '../command/cliCommand';
import {PreconditionError} from '../errors/preconditionError';
import {DecoratorFunction} from '../decorators/decoratorFunction';

const thrower = (reason: string) => {
  throw new PreconditionError(`${reason} Precondition Error`);
};

export const mockPreconditions = <
  PreconditionStatus extends Record<string, boolean>
>(
  preconditionStatus: PreconditionStatus
) => {
  type PreconditionKeys = keyof PreconditionStatus & string;
  type PreconditionPromises = Record<PreconditionKeys, DecoratorFunction>;

  const keys: PreconditionKeys[] = Object.keys(preconditionStatus);
  const mockedPreconditions: Partial<PreconditionPromises> = {};
  for (const key of keys) {
    mockedPreconditions[key] = (_target?: CLICommand) =>
      new Promise<void>((resolve) =>
        preconditionStatus[key] ? resolve() : thrower(key)
      );
  }

  return mockedPreconditions as PreconditionPromises;
};
