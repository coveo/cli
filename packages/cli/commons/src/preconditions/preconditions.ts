import {Before} from '../decorators/before';
import { DecoratorFunction } from '../decorators/decoratorFunction';

/**
 * @deprecated use @type {DecoratorFunction}
 */
export type PreconditionFunction = DecoratorFunction;

/**
 * @deprecated use @function {Before}
 */
export const Preconditions = Before;
