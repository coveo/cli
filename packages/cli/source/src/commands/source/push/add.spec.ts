jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/preconditions/authenticated');
jest.mock('@coveo/cli-commons/preconditions/apiKeyPrivilege');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('@coveo/push-api-client');

import {test} from '@oclif/test';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {PushSource} from '@coveo/push-api-client';
import {APIError} from '@coveo/cli-commons/errors/apiError';
import {
  BatchUploadDocumentsError,
  BatchUploadDocumentsSuccess,
} from '../../../lib/__testsUtils__/batchUploadDocumentsFromFilesReturn';
import {
  HasNecessaryCoveoPrivileges,
  IsAuthenticated,
} from '@coveo/cli-commons/preconditions';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';
import {formatCliLog} from '@coveo/cli-commons-dev/testUtils/jestSnapshotUtils';

describe('source:push:add', () => {
  const mockSetSourceStatus = jest.fn();
  const mockBatchUpdate = jest.fn();
  const mockedClient = jest.mocked(AuthenticatedClient);
  const mockedSource = jest.mocked(PushSource);
  const mockedIsAuthenticated = jest.mocked(IsAuthenticated);
  const mockedHasNecessaryCoveoPrivileges = jest.mocked(
    HasNecessaryCoveoPrivileges
  );
  const doMockPreconditions = function () {
    const preconditionStatus = {
      authentication: true,
      privileges: true,
    };
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
    mockedHasNecessaryCoveoPrivileges.mockReturnValue(
      mockedPreconditions.privileges
    );
  };

  const doMockSuccessBatchUpload = (numberOfDocuments = 2) => {
    mockBatchUpdate.mockReturnValue(
      new BatchUploadDocumentsSuccess(numberOfDocuments)
    );
  };

  const doMockErrorBatchUpload = () => {
    mockBatchUpdate.mockReturnValue(new BatchUploadDocumentsError());
  };
  const doMockAuthenticatedClient = () => {
    mockedClient.mockImplementation(
      () =>
        ({
          cfg: {
            get: () => ({
              accessToken: 'the_token',
              organization: 'the_org',
              region: 'au',
              environment: 'prod',
            }),
          },
        } as unknown as AuthenticatedClient)
    );
  };

  const doMockSource = () => {
    mockedSource.mockImplementation(
      () =>
        ({
          batchUpdateDocumentsFromFiles: mockBatchUpdate,
          setSourceStatus: mockSetSourceStatus,
        } as unknown as PushSource)
    );
  };

  beforeEach(() => {
    doMockAuthenticatedClient();
    doMockSource();
    doMockPreconditions();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('when the batch upload is successful', () => {
    beforeEach(() => {
      doMockSuccessBatchUpload();
    });

    test
      .stdout()
      .stderr()
      .command(['source:push:add', 'mysource'])
      .catch((err) => {
        expect(err.message).toMatchSnapshot();
      })
      .it('throws when no flags are specified');

    test
      .stdout()
      .stderr()
      .command([
        'source:push:add',
        'mysource',
        '-f',
        'foo.json',
        '-d',
        'directory',
      ])
      .it('should accept files and folder within the same command', () => {
        expect(mockBatchUpdate).toHaveBeenCalledWith(
          expect.anything(),
          ['directory', 'foo.json'],
          expect.anything()
        );
      });

    test
      .stdout()
      .stderr()
      .command(['source:push:add', 'mysource', '-f', 'whatever'])
      .it('pass correct configuration information to push-api-client', () => {
        expect(mockedSource).toHaveBeenCalledWith('the_token', 'the_org', {
          environment: 'prod',
          region: 'au',
        });
      });

    test
      .stdout()
      .stderr()
      .command(['source:push:add', 'mysource', '-f', 'somepath'])
      .it('should create missing fields by default', () => {
        expect(mockBatchUpdate).toHaveBeenCalledWith(
          expect.anything(),
          expect.anything(),
          expect.objectContaining({createFields: true})
        );
      });

    test
      .stdout()
      .stderr()
      .command([
        'source:push:add',
        'mysource',
        '--no-createMissingFields',
        '-f',
        'somepath',
      ])
      .it('should skip field creation if specified', () => {
        expect(mockBatchUpdate).toHaveBeenCalledWith(
          expect.anything(),
          expect.anything(),
          expect.objectContaining({createFields: false})
        );
      });

    describe('when using the -d flag', () => {
      test
        .stdout()
        .stderr()
        .command(['source:push:add', 'mysource', '-d', 'somepath'])
        .it(
          'should output feedback message when uploading documents',
          (ctx) => {
            expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
          }
        );
      test
        .stdout()
        .stderr()
        .command(['source:push:add', 'mysource', '-d', 'somepath'])
        .it('should show deprecated flag warning', (ctx) => {
          expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
        });
    });

    describe.each([1, 2, 3, 4, 5, 6])(
      'when uploading %d documents',
      (numberOfDocuments) => {
        beforeEach(() => {
          doMockSuccessBatchUpload(numberOfDocuments);
        });

        test
          .stdout()
          .stderr()
          .command(['source:push:add', 'mysource', '-f', 'somepath'])
          .it('should print the appropriate # of documents', (ctx) => {
            expect(formatCliLog(ctx.stdout)).toMatchSnapshot();
          });
      }
    );

    test
      .stdout()
      .stderr()
      .command(['source:push:add', 'mysource', '-f', 'somepath'])
      .it('should update the source status', () => {
        expect(mockSetSourceStatus).toHaveBeenNthCalledWith(
          1,
          'mysource',
          'REFRESH'
        );
        expect(mockSetSourceStatus).toHaveBeenNthCalledWith(
          2,
          'mysource',
          'IDLE'
        );
      });
  });

  describe('when the batch upload fails', () => {
    beforeEach(() => {
      doMockErrorBatchUpload();
    });

    test
      .stdout()
      .stderr()
      .command(['source:push:add', 'mysource', '-f', 'somepath'])
      .catch((err) => {
        expect(err).toBeInstanceOf(APIError);
        expect(err.message).toMatchSnapshot();
      })
      .it('returns an information message on add failure from the API');
  });
});
