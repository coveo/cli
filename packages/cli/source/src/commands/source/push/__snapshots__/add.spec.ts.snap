// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source:push:add when the batch upload fails returns an information message on add failure from the API 1`] = `
"Error while trying to add document.

Status code: 412
Error code: [31mBAD_REQUEST[39m
Message: [31mthis is a bad request and you should feel bad[39m"
`;

exports[`source:push:add when the batch upload is successful throws when no flags are specified 1`] = `"You must set the \`files\` flag. Use \`source:push:add --help\` to get more information."`;

exports[`source:push:add when the batch upload is successful when uploading 1 documents should print the appropriate # of documents 1`] = `
"Success: 1 document accepted by the Push API from filename0.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when uploading 2 documents should print the appropriate # of documents 1`] = `
"Success: 2 documents accepted by the Push API from filename0, filename1.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when uploading 3 documents should print the appropriate # of documents 1`] = `
"Success: 3 documents accepted by the Push API from filename0, filename1, filename2.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when uploading 4 documents should print the appropriate # of documents 1`] = `
"Success: 4 documents accepted by the Push API from filename0, filename1, filename2, filename3.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when uploading 5 documents should print the appropriate # of documents 1`] = `
"Success: 5 documents accepted by the Push API from filename0, filename1, filename2, filename3, filename4.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when uploading 6 documents should print the appropriate # of documents 1`] = `
"Success: 6 documents accepted by the Push API from filename0, filename1, filename2, filename3, filename4 and 1 more ....Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when using the -d flag should output feedback message when uploading documents 1`] = `
"  console.error
     â€º   Warning: Use the \`files\` flag instead

      142 |     const {flags} = await this.parse(SourcePushAdd);
      143 |     if (flags.file || flags.folder) {
    > 144 |       CliUx.ux.warn('Use the \`files\` flag instead');
          |                ^
      145 |     }
      146 |   }
      147 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at SourcePushAdd.warn [as showDeprecatedFlagWarning] (src/commands/source/push/add.ts:144:16)
      at SourcePushAdd.run (src/commands/source/push/add.ts:76:5)
      at SourcePushAdd._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

Success: 2 documents accepted by the Push API from filename0, filename1.Status code: 202 ðŸ‘Œ
    
"
`;

exports[`source:push:add when the batch upload is successful when using the -d flag should show deprecated flag warning 1`] = `
"  console.error
     â€º   Warning: Use the \`files\` flag instead

      142 |     const {flags} = await this.parse(SourcePushAdd);
      143 |     if (flags.file || flags.folder) {
    > 144 |       CliUx.ux.warn('Use the \`files\` flag instead');
          |                ^
      145 |     }
      146 |   }
      147 |

      at Object.warn (../../../node_modules/@oclif/core/lib/errors/index.js:59:13)
      at SourcePushAdd.warn [as showDeprecatedFlagWarning] (src/commands/source/push/add.ts:144:16)
      at SourcePushAdd.run (src/commands/source/push/add.ts:76:5)
      at SourcePushAdd._run (../../../node_modules/@oclif/core/lib/command.js:81:22)
      at Config.runCommand (../../../node_modules/@oclif/core/lib/config/config.js:272:25)
      at Object.run (../../../node_modules/@oclif/test/lib/command.js:21:13)
      at Object.run (../../../node_modules/fancy-test/lib/base.js:68:25)

Success: 2 documents accepted by the Push API from filename0, filename1.Status code: 202 ðŸ‘Œ
    
"
`;
